practice/tic-tac-toe/dict_board.py

```diff
@@ -72,29 +69,47 @@ def __str__(self):
         O| |O
          | |O
         """
-        row_1 = [' ', ' ', ' ']
-        row_2 = [' ', ' ', ' ']
-        row_3 = [' ', ' ', ' ']
+        print_list = [' ', ' ', ' '], [' ', ' ', ' '], [' ', ' ', ' ']
         i = 0
-        for item in self._tokens_to_coords:
-            for key, value in self._tokens_to_coords[i].items():
-                for coord in value:
-                    x = coord[0]
-                    y = coord[1]
-                    if y == 0:
-                        row_1[x] = key
-                    if y == 1:
-                        row_2[x] = key
-                    if y == 2:
-                        row_3[x] = key
-            i += 1
-        row_1 = '|'.join(row_1)
-        row_2 = '|'.join(row_2)
-        row_3 = '|'.join(row_3)
-        return str(row_1 + '\n' + row_2 + '\n' + row_3)
+        for token in self._tokens_to_coords:
+                for key, value in self._tokens_to_coords[i].items():
+                    for coord in value:
+                        x = coord[0]
+                        y = coord[1]
+                        print_list[y][x] = key
+                i += 1
+        joined_rows = ['|'.join(row) for row in print_list]
+        return '\n'.join(joined_rows)
 
     def _append_to_tokens_to_coords(self, key, coords):
         group = key
         if group not in self._tokens_to_coords:
             self._tokens_to_coords[group] = []
         self._tokens_to_coords[group].append(coords)
+
+
+def _get_winning_combinations():
+    winning_combinations = []
+    for i in range(8):
+        winning_combinations.append([])
+    x = 0
+    for i in range(3):
+        for y in range(3):
+            winning_combinations[i].append((x, y))
+        x += 1
+    y = 0
+    for i in range(3, 6):
+        for x in range(3):
+            winning_combinations[i].append((x, y))
+        y += 1
+    for i in range(6, 7):
+        for x in range(3):
+            winning_combinations[i].append((x, x))
+    x = 3
+    y = 0
+    for i in range(7, 8):
+        for x in reversed(range(x)):
+            winning_combinations[i].append((x, y))
+            y += 1
+    return winning_combinations
```

Comment:
If this problem is broken apart into generating each kind of wining combination separately, it becomes very easy to phrase as list comprehensions.

```py
WINNING_COL_COMBINATIONS = [[(x, y) for y in range(3)] for x in range(3)]
WINNING_ROW_COMBINATIONS = [[(x, y) for x in range(3)] for y in range(3)]
WINNING_DIAG_COMBINATIONS = [(i, i) for i in range(3)] + [(3 - i, i) for i in range(3)]
WINNING_COMBINATIONS = WINNING_COL_COMBINATIONS + WINNING_ROW_COMBINATIONS + WINNING_DIAG_COMBINATIONS
```